### PR TITLE
Support SharedLister for NodeInfo

### LINKED ISSUE #70: cache Node data via single unified Node data
The Pod is cached in one scheduling cycle.
https://github.com/kubernetes-sigs/kube-scheduler-wasm-extension/blob/main/guest/internal/cyclestate/cyclestate.go#L17-L19

And we can do the similar with Nodes.

There are currently two types for Nodes.
- nodeList in PreScore https://github.com/kubernetes-sigs/kube-scheduler-wasm-extension/blob/main/guest/prescore/prescore.go#L93
- nodeInfo in Filter https://github.com/kubernetes-sigs/kube-scheduler-wasm-extension/blob/main/guest/filter/filter.go#L81

We can unify them and cache Nodes as well like a Pod.

/kind feature

### PR DESCRIPTION
**Pull request by sanposhiho**
#### What type of PR is this?

/kind feature
/kind cleanup

#### What this PR does / why we need it:

- Improve how the guest caches nodes for a better efficiency.
  - And, wire SharedLister().NodeInfos() and cached Nodes.
- Support SharedLister for NodeInfo.
  - Only the basic node's data and `ImageStates` is supported for now. We can support other fields later.
- Add [ImageLocality plugin](https://github.com/kubernetes/kubernetes/tree/master/pkg/scheduler/framework/plugins/imagelocality), ported from the in-tree plugin.

#### Which issue(s) this PR fixes:

Fixes #70 
Part of #73

#### Does this PR introduce a user-facing change?

```release-note
SharedLister for NodeInfo is supported.
```

#### What are the benchmark results of this change?

```benchstat
                                       │  before.txt  │              after.txt               │
                                       │    sec/op    │    sec/op      vs base               │
Example_NodeNumber/Simple/New-12         563.2m ±  2%    584.6m ±  3%   +3.82% (p=0.015 n=6)
Example_NodeNumber/Simple/Run-12         93.41µ ± 26%   108.65µ ± 12%        ~ (p=0.093 n=6)
Example_NodeNumber/Simple_Log/New-12     552.2m ±  2%    614.9m ± 16%  +11.35% (p=0.002 n=6)
Example_NodeNumber/Simple_Log/Run-12     109.4µ ±  8%    108.9µ ± 28%        ~ (p=0.485 n=6)
Example_NodeNumber/Advanced/New-12       633.0m ±  3%    660.1m ±  3%   +4.28% (p=0.002 n=6)
Example_NodeNumber/Advanced/Run-12       37.80µ ± 25%    38.30µ ±  4%        ~ (p=0.589 n=6)
Example_NodeNumber/Advanced_Log/New-12   649.9m ± 12%    648.1m ±  5%        ~ (p=0.937 n=6)
Example_NodeNumber/Advanced_Log/Run-12   42.80µ ±  3%    44.93µ ± 10%   +4.98% (p=0.002 n=6)
geomean                                  6.175m          6.486m         +5.03%
```

### CODE REVIEW

File: guest/api/framework.go
Comment (Gekko0114) [inline guest/api/framework.go:8]: minor comment:
I don't think int64 is necessary for this field. How about using int32 or a smaller variable type?

File: guest/handle/sharedlister/api/sharedlister.go
Comment (Gekko0114) [inline guest/handle/sharedlister/api/sharedlister.go:8]: minor:
How about adding the comment that StorageInfos are not implemented yet?

File: guest/internal/prefilter/prefilter_exports.go
Comment (Gekko0114) [inline guest/internal/prefilter/prefilter_exports.go:53]: Why does this `isFullNodeInfoList` logic work? Is it because the Node used in the scheduler's step always goes through the prefilter process?

File: guest/prescore/prescore.go
Comment (Gekko0114) [inline guest/prescore/prescore.go:94]: Why do you define filteredNodeInfoList even though sharedLister has nodeinfos? Is this because wasm's prescore step is sometimes called without calling previous steps like prefilter?

**Review by Gekko0114 — COMMENTED** [review]
Thanks!
Though I haven't finished reading all the code, I left several minor comments and questions :)

File: guest/prescore/prescore.go
Comment (sanposhiho) [reply guest/prescore/prescore.go:94]: Because, as documented at this line, we have to know which nodes have gone thru the filtering phase.

As you can see in `filteredNodeInfoList.lazyItems`, actually `filteredNodeInfoList` utilizes nodes in `sharedLister` too to use the cached nodes there. 
Also, we don't want to put this "listing nodes after filtering" logic into `sharedLister` because we want to keep `sharedLister` pure, and "listing nodes that have gone thru the filtering phase" is specific to PreScore.

File: guest/api/framework.go
Comment (sanposhiho) [reply guest/api/framework.go:4]: > Nodes field is not necessary for now?

Yes, we don't use it (actually nothing uses it in upstream too). 

> How about leaving the comment why this field is not implemented?

I don't want to leave comments on everything unsupported. (We have too many unsupported things.

File: guest/handle/sharedlister/api/sharedlister.go
Comment (sanposhiho) [reply guest/handle/sharedlister/api/sharedlister.go:8]: Ditto, I don't want to leave comments on everything unsupported.

File: guest/internal/prefilter/prefilter_exports.go
Comment (sanposhiho) [reply guest/internal/prefilter/prefilter_exports.go:53]: Yes, we call preFilter regardless of whether an user implements preFilter or not, so that we can reset all stuff the start of the scheduling cycle.

**Comment by sanposhiho** [comment]
@Gekko0114  Addressed or replied to your comments.

File: scheduler/plugin/host.go
Comment (Gekko0114) [inline scheduler/plugin/host.go:202]: Sorry, but at first, I couldn't understand this code.
This code retrieves the node name from the wasm module and returns nodeinfo to the wasm module, right?
Additionally, I couldn't understand why this function is necessary even though k8sSchedulerCurrentNodeNameFn exists. (I understand it now)

Could you leave a brief comment explaining this function?

**Review by Gekko0114 — APPROVED** [review]
LGTM, but leaves lgtm label for another reviewer.
Also left several minor comments.

File: examples/imagelocality/image_locality.go
Comment (Gekko0114) [reply examples/imagelocality/image_locality.go:14]: Oh sorry, I didn't this expression, but this doesn't seem typo. Please ignore this comment

File: scheduler/plugin/host.go
Comment (sanposhiho) [reply scheduler/plugin/host.go:203]: Added the comment though, what's unclear?

Comment (Gekko0114) [reply scheduler/plugin/host.go:203]: Lgtm 😀

**Comment by sanposhiho** [comment]
Merging the PR. 

@Gekko0114 (fyi @utam0k ) 
This repo is under development anyway and doesn't need to be conservative. So, let's not wait for other people's approval like k/k.
I mean not only PRs from me, but also ones from anyone. I know people besides me cannot `/approve` though, I'll basically just put the approval for PRs that either of you put the stamp.

**Pull request merged**
